modules/standard/Version.chpl

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright 2020 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+
+This module contains features that support compile-time reasoning
+about version numbers in general, and Chapel version numbers
+specifically.  In more detail, it features:
+
+* :var:`chplVersion`: the version number of the copy of ``chpl`` used
+  to compile the program.
+
+* two types that can be used to represent version numbers:
+
+  * :type:`semanticVersion` which represents a semantic version value
+
+  * :type:`sourceVersion` which is a semantic version plus a commit value
+
+* :proc:`version`: a utility function for creating new version values
+
+Version types in this module are defined in terms of `param` values to
+support compile-time reasoning about versions and code specialization.
+
+The version types support:
+
+* being printed out or cast to a ``param`` string
+
+* compile-time comparisons between version values via ``==``, ``!=``,
+  ``<``, ``<=``, ``>``, and ``>``, including mixed types.  Generally
+  speaking, "less than" corresponds to "is an earlier version than."
+
```

Comment:
It'd be nice to have a code example here showing how to compare something like `chplVersion > version(1, 22)`, but I'm struggling to come up with an example that isn't too artificial and won't become out of date each release.

Reply:
What do you think about:

```chapel
    if chplVersion < version(2,0) then                                          
      writeln("Warning: Using a not-yet-stable version of Chapel");             
```

Reply:
Or maybe instead of `writeln()`, `compilerWarning()` to subtly emphasize that it's a compile-time folded conditional?

Reply:
Yeah, I like the compiler warning

Reply:
Thinking about this more over lunch, an alternative that might be less revealing of Chapel's stability would be something like:

```chapel
if chplVersion < version(1,22) then
  compilerError("Arkouda is intended to be used with Chapel 1.22 or later");
```

(of course, this conditional by definition will never be false, but we don't have to admit to that... :) ).

The only slightly weird thing about this is that it mentions Arkouda in our docs.  Could say something vaguer like "This code..." instead if that seemed weird.